pkg/loadbalancer/services.go

```diff
@@ -0,0 +1,415 @@
+		return err
+	}
+
+	// Bump the backend revision for each of the referenced services to force
+	// reconciliation.
+	revision := s.bes.Revision(txn)
+	for _, name := range be.ReferencedBy.AsSlice() {
+		iter, _ := s.svcs.Get(txn, ServiceNameIndex.Query(name))
+		for svc, _, ok := iter.Next(); ok; svc, _, ok = iter.Next() {
+			svc = svc.Clone()
+			svc.BackendRevision = revision
+			svc.Status = reconciler.StatusPending()
+			if _, _, err := s.svcs.Insert(txn, svc); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Services) DebugDump(txn statedb.ReadTxn) {
+	w := tabwriter.NewWriter(os.Stdout, 5, 0, 3, ' ', 0)
+
+	fmt.Fprintln(w, "--- Services ---")
+	fmt.Fprintln(w, strings.Join((*Service)(nil).TableHeader(), "\t"))
+	iter, _ := s.svcs.All(txn)
+	for svc, _, ok := iter.Next(); ok; svc, _, ok = iter.Next() {
+		fmt.Fprintln(w, strings.Join(svc.TableRow(), "\t"))
+	}
+
+	fmt.Fprintln(w, "--- Backends ---")
+	fmt.Fprintln(w, strings.Join((*ServiceBackend)(nil).TableHeader(), "\t"))
+	iterBe, _ := s.bes.All(txn)
+	for be, _, ok := iterBe.Next(); ok; be, _, ok = iterBe.Next() {
+		fmt.Fprintln(w, strings.Join(be.TableRow(), "\t"))
+	}
+
+	w.Flush()
+}
+
+// UnsafeRWTable returns the RWTable[*Service] for direct unvalidated write access. Only use this
+// for reconcilers.
+func (s *Services) UnsafeRWTable() statedb.RWTable[*Service] {
+	return s.svcs
+}
+
+func (s *Services) Wrapper(orig svcManager) svcManager {
+	if !s.IsEnabled() {
+		return orig
+	}
+	return &NewServicesWrapper{s.db, orig, s}
+}
+
+// svcManager is a clone of watchers.svcManager
+type svcManager interface {
+	DeleteService(frontend L3n4Addr) (bool, error)
+	GetDeepCopyServiceByFrontend(frontend L3n4Addr) (*SVC, bool)
+	UpsertService(*SVC) (bool, ID, error)
+}
+
+// NewServicesWrapper implements enough of ServiceManager to work as a replacement in pkg/k8s/watchers.
+// This is not efficient at all since we do one service upsert/delete per transaction.
+//
+// This wrapper ignores the IDs.
+type NewServicesWrapper struct {
+	db   *statedb.DB
+	orig svcManager
+	s    *Services
+}
+
+// DeleteService implements svcManager.
+func (n *NewServicesWrapper) DeleteService(frontend L3n4Addr) (bool, error) {
+	wtxn := n.s.WriteTxn()
+	defer wtxn.Commit()
+	svc, _, found := n.s.Services().First(wtxn, ServiceL3n4AddrIndex.Query(frontend))
+	if !found {
+		return false, nil
+	}
+	old, err := n.s.DeleteService(wtxn, svc.Name, frontend)
+	return old != nil, err
+}
+
+// GetDeepCopyServiceByFrontend implements svcManager.
+func (n *NewServicesWrapper) GetDeepCopyServiceByFrontend(frontend L3n4Addr) (*SVC, bool) {
+	return n.orig.GetDeepCopyServiceByFrontend(frontend)
+}
+
+// UpsertService implements svcManager.
+func (n *NewServicesWrapper) UpsertService(svc *SVC) (bool, ID, error) {
+	wtxn := n.s.WriteTxn()
+	defer wtxn.Abort()
+
+	params := ServiceParams{
+		Name:                      svc.Name,
+		Frontend:                  svc.Frontend.L3n4Addr,
```

Comment:
Yes, there's a separate service for each type, port, address. I'm starting to lean towards calling this a `Frontend` instead of `Service` as it's confusing when this is not semantically the same as the K8s one.

Reply:
Renamed `Service` to `Frontend`